tests/virt_autotest/sriov_pci_passthrough.pm

```diff
@@ -0,0 +1,381 @@
+  my ($vm, $host_device_id) = @_;
+
+  #get neccessary device config from host
+  my $host_device_xml = script_output "virsh nodedev-dumpxml $host_device_id";
+  $host_device_xml =~ /\<domain\>(\w+)\<\/domain\>.*\<bus\>(\w+)\<\/bus\>.*\<slot\>(\w+)\<\/slot\>.*\<function\>(\w+)\<\/function\>/s;
+  my ($dev_domain, $dev_bus, $dev_slot, $dev_func) = ($1, $2, $3, $4);
+  
+  #'virsh attach-interface' can plug the network device to guest with the BDF in commandline without a seperate device xml file
+  #but the SUSE document use 'virsh attach-device', so the test follows it
+  #create the device xml to passthrough to vm
+  my $vf_host_addr_xml = "<address type='pci' domain='$dev_domain' bus='$dev_bus' slot='$dev_slot' function='$dev_func'/>";
+  assert_script_run("echo \"<interface type='hostdev'>\n  <source>\n    $vf_host_addr_xml\n  </source>\n</interface>\" > $host_device_id.xml", 60);
+  upload_logs("$host_device_id.xml");
+
+  #attach device to vm
+  assert_script_run "virsh -d 1 attach-device $vm $host_device_id.xml --persistent";
+  save_screenshot;
+
+  #get the mac address and bdf by parsing the domain xml
+  #tips: there may be multiple interfaces and multiple hostdev devices in the guest
+  my %nic;
+  my $nics_count = script_output "virsh dumpxml $vm | grep -c \"<interface.*type='hostdev'\"";
+  my $devs_xml = script_output "virsh dumpxml $vm | sed -n \"/<interface.*type='hostdev'/,/<\\/devices/p\"";
+  $host_device_id =~ /pci_([a-z\d]+)_([a-z\d]+)_([a-z\d]+)_([a-z\d]+)/;
+  my ($dom, $bus, $slot, $func) = ($1, $2, $3, $4);  #these are different with those in host_devices.xml 
+  for (my $i=0; $i<$nics_count; $i++) {
+    $devs_xml =~ /\<interface.*?type='hostdev'.*?\<\/interface\>/s;    #minimal match. 
+    my $nic_xml = $&;
+    if ($nic_xml =~ /domain='(0x)?$dom' +bus='(0x)?$bus' +slot='(0x)?$slot' +function='(0x)?$func'/){
+
+      #get the vf xml in vm config file for later unplugging to use
+      $nic{xml} = $nic_xml;
+
+      #get mac address
+      $nic_xml =~ /\<mac.*address='(.*)'/;
+      $nic{mac} = $1;
+
+      #get bdf for guests on KVM
+      $nic_xml =~ s/\<source.*\<\/source\>//s;
+      if ($nic_xml =~ /bus='(\w+)'.*slot='(\w+)'.*function='(\w+)'/) {
+        my ($bus, $slot, $func) = ($1, $2, $3);
+        $nic{bdf} = $bus . ":" . $slot . "." . $func;
+      }
+      else {
+        #have to get bdf by other means for guests on XEN
+        #but no these file on pv guest on XEN, need to find another way!! ???
+        $nic{bdf} = script_output "ssh root\@$vm \"if [ -e /sys/devices/pci-0/pci????:?? ]; then grep -H '$nic{mac}' /sys/devices/pci-0/*/*/net/*/address | cut -d '/' -f6; else grep -H '$nic{mac}' /sys/devices/*/*/net/*/address | cut -d '/' -f5; fi\"";
+#        $nic{bdf} = script_output "ssh root\@$vm \"grep -H '$nic{mac}' /sys/devices/*/*/net/*/address | cut -d '/' -f5\"";
+      }
+      last;
+
+    }
+    else {
+      $devs_xml =~ s/\<interface.*?type='hostdev'.*?\<\/interface\>//s;  
+    } 
+  }
```

Comment:
This is the only way to get mac and bdf ?

Reply:
Possible to use xmlstarlet command line tool ? This one is very capable of query xml file.

Reply:
I do not know if it is the only way, but it is the only way I found to determine the relation between host_vf_slot and its_mac_in_vm.

I do know little about xmlstarlet, but I can do some research. thanks.

Reply:
I can help provide example to use xmlstarlet in this scenario if you need.

Reply:
Thanks. I will consider to use xmlstarlet when later I am going to deploy the test in openqa in OSD. But for now, I have to use the current logic as it works and the priority is not high.